### PR TITLE
Refs #29908 -- Optimized known related objects assignment.

### PR DESCRIPTION
**Pull request by charettes**
Since [CPython implements a C level attrgetter(*attrs)](https://github.com/python/cpython/blob/ad65f15581173542f1d2a9968a63bee272510ce3/Modules/_operator.c#L1109-L1245) it even outperforms the most common case of a single known related object since the resulting attribute values tuple is built in C.

Note that this also restores the `_known_related_objects` key types for relationships involving a single target field.

### CODE REVIEW

File: django/db/models/fields/related_descriptors.py [hunk 1]
```diff
@@ -581,10 +582,17 @@ def _apply_rel_filters(self, queryset):
...
-                rel_obj_id = tuple([
-                    getattr(self.instance, target_field.attname)
-                    for target_field in self.field.get_path_info()[-1].target_fields
-                ])
+                try:
+                    target_field = self.field.target_field
+                except FieldError:
+                    # The relationship has multiple target fields, use a tuple
+                    # for related object id.
+                    rel_obj_id = tuple(
```
Comment (timgraham) [inline django/db/models/fields/related_descriptors.py]: Doesn't `tuple([])` perform better?

File: django/db/models/fields/related_descriptors.py [hunk 2]
```diff
@@ -581,10 +582,17 @@ def _apply_rel_filters(self, queryset):
...
                 # one-to-many relationships instead and break known related
                 # objects assignment.
-                rel_obj_id = tuple([
-                    getattr(self.instance, target_field.attname)
-                    for target_field in self.field.get_path_info()[-1].target_fields
-                ])
+                try:
+                    target_field = self.field.target_field
+                except FieldError:
+                    # The relationship has multiple target fields, use a tuple
```
Comment (timgraham) [inline django/db/models/fields/related_descriptors.py]: I'd change the comma to a period.

File: django/db/models/fields/related_descriptors.py (hunk 1 above)
Comment (charettes) [reply django/db/models/fields/related_descriptors.py]: Supposedly but given this wasn't in a hot path I figured this form was more readable. I'll add back the list comprehension wrapping.

File: django/db/models/fields/related_descriptors.py (hunk 2 above)
Comment (charettes) [reply django/db/models/fields/related_descriptors.py]: Will do.

**Pull request merged**